clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -134,22 +134,35 @@ public Fetcher(ConsumerNetworkClient client,
      */
     public void sendFetches() {
         for (Map.Entry<Node, FetchRequest> fetchEntry: createFetchRequests().entrySet()) {
+            final Node fetchTarget = fetchEntry.getKey();
             final FetchRequest fetch = fetchEntry.getValue();
             client.send(fetchEntry.getKey(), ApiKeys.FETCH, fetch)
                     .addListener(new RequestFutureListener<ClientResponse>() {
                         @Override
-                        public void onSuccess(ClientResponse response) {
-                            handleFetchResponse(response, fetch);
+                        public void onSuccess(ClientResponse resp) {
+                            synchronized (completedFetches) {
+                                sensors.fetchLatency.record(resp.requestLatencyMs());
+                                FetchResponse response = new FetchResponse(resp.responseBody());
+                                completedFetches.add(new FetchCompletion(fetch, response));
+                            }
                         }
 
                         @Override
                         public void onFailure(RuntimeException e) {
-                            log.debug("Fetch failed", e);
+                            log.debug("Fetch request to {} failed", fetchTarget, e);
                         }
                     });
         }
     }
 
+    private void drainCompletedFetches() {
+        synchronized (completedFetches) {
+            for (FetchCompletion completion : completedFetches)
+                handleFetchResponse(completion.request, completion.response);
```

Comment:
You probably want to remove these as you process them so that if you have multiple completed fetches and one of them throws an exception during parsing you won't reprocess the ones before it. (Given the other bug identified on the mailing list, we should probably have a test for this if possible)